src/Native/LibTorchSharp/THSNN.cpp

```diff
@@ -986,6 +1020,185 @@ Tensor THSNN_FeatureAlphaDropout_forward(const NNModule module, const Tensor ten
+{
+    CATCH_RETURN_NNModule(
+        auto opts = torch::nn::PixelUnshuffleOptions(downscale_factor);
+        res = create_module<torch::nn::PixelUnshuffleImpl>(opts, outAsAnyModule);
+    );
+}
+
+Tensor   THSNN_PixelUnshuffle_forward(const NNModule module, const Tensor tensor)
+{
+    CATCH_TENSOR((*module)->as<torch::nn::PixelUnshuffle>()->forward(*tensor));
+}
+
+template<typename T>
+void ApplyUpsampleMode(T& opts, const int8_t mode)
+{
+    if (mode == 0)
+        opts = opts.mode(torch::kNearest);
+    if (mode == 1)
+        opts = opts.mode(torch::kLinear);
+    if (mode == 2)
+        opts = opts.mode(torch::kBilinear);
+    if (mode == 3)
+        opts = opts.mode(torch::kBicubic);
+    if (mode == 4)
+        opts = opts.mode(torch::kTrilinear);
+}
+
+template<typename T>
+void ApplyInterpolateMode(T& opts, const int8_t mode)
+{
+    if (mode == 0)
+        opts = opts.mode(torch::kNearest);
+    if (mode == 1)
+        opts = opts.mode(torch::kLinear);
+    if (mode == 2)
+        opts = opts.mode(torch::kBilinear);
+    if (mode == 3)
+        opts = opts.mode(torch::kBicubic);
+    if (mode == 4)
+        opts = opts.mode(torch::kTrilinear);
+    if (mode == 5)
+        opts = opts.mode(torch::kArea);
+}
+
+NNModule THSNN_Upsample_ctor(const int64_t* size, const int size_len, const double* scale_factor, const int scale_factor_len, const int8_t mode, const int8_t align_corners, NNAnyModule* outAsAnyModule)
+{
+    auto opts = torch::nn::UpsampleOptions();
+    // align_corners -- 0=None, 1=true, 2=false
+    if (align_corners != 0)
+        opts.align_corners(align_corners == 1);
+    ApplyUpsampleMode(opts, mode);
+
+    CATCH_RETURN_NNModule(
+        if (size_len > 0) {
+            std::vector<int64_t> sizes;
+            for (int i = 0; i < size_len; ++i) {
+                sizes.push_back(size[i]);
+            }
+            opts.size(sizes);
+        }
+        if (scale_factor_len > 0) {
+            std::vector<double> scales;
+            for (int i = 0; i < scale_factor_len; ++i) {
+                scales.push_back(scale_factor[i]);
+            }
+            opts.scale_factor(scales);
+        }
+        res = create_module<torch::nn::UpsampleImpl>(opts, outAsAnyModule);
+    );
+}
+
+Tensor   THSNN_Upsample_forward(const NNModule module, const Tensor tensor)
+{
+    CATCH_TENSOR((*module)->as<torch::nn::Upsample>()->forward(*tensor));
+}
+
+template<typename T>
+void ApplyPadMode(T& opts, const int64_t padding)
+{
+    if (padding == 1)
+        opts = opts.mode(torch::kReflect);
+    if (padding == 2)
+        opts = opts.mode(torch::kReplicate);
+    if (padding == 3)
+        opts = opts.mode(torch::kCircular);
+    if (padding == 4)
+        opts = opts.mode(torch::kConstant);
+}
+
+template<typename T>
+void ApplyGridMode(T& opts, const int8_t mode)
+{
+    if (mode == 0)
+        opts = opts.mode(torch::kNearest);
+    if (mode == 2)
+        opts = opts.mode(torch::kBilinear);
+    //if (mode == 3)
```

Comment:
what's up here?

Reply:
With the commented line? The PyTorch docs say that it should be supported, but the C++ variant<> type declaration doesn't allow it. I left it in for future use.

Reply:
I'll put a comment in the code to clarify.